parity-clib/src/lib.rs

```diff
@@ -255,26 +214,70 @@ pub unsafe extern fn parity_unsubscribe_ws(session: *const c_void) {
 }
 
 #[no_mangle]
-pub unsafe extern fn parity_set_panic_hook(callback: Callback, param: *mut c_void) {
+pub extern fn parity_set_panic_hook(callback: CCallback, param: *mut c_void) {
 	let cb = CallbackStr {user_data: param, function: callback};
 	panic_hook::set_with(move |panic_msg| {
-		cb.call(panic_msg.as_bytes());
+		cb.call(panic_msg);
 	});
 }
 
-// Internal structure for handling callbacks that get passed a string.
-struct CallbackStr {
-	user_data: *mut c_void,
-	function: Callback,
+// WebSocket event loop
+fn parity_ws_worker(client: &RunningClient, query: &str, callback: Arc<Callback>) -> *const c_void {
+	let (tx, mut rx) = mpsc::channel(1);
+	let session = Arc::new(PubSubSession::new(tx));
+	let query_future = client.rpc_query(query, Some(session.clone()));
+	let weak_session = Arc::downgrade(&session);
+	let _handle = thread::Builder::new()
+		.name("ws-subscriber".into())
+		.spawn(move || {
+			// Wait for subscription ID
+			// Note this may block forever and be can't destroyed using the session object
+			// However, this will likely timeout or be catched the RPC layer
+			if let Ok(Some(response)) = query_future.wait() {
+				callback.call(&response);
+			} else {
+				callback.call(error::SUBSCRIBE);
+				return;
+			}
+
+			while weak_session.upgrade().map_or(0, |session| Arc::strong_count(&session)) > 1 {
```

Comment:
This is not equivalent to the loop used before if the condition returns false at the first iteration (is that ok?).

Reply:
Nice catch

Yeah, that is ok that means that the `sink` has been dropped and we should terminate. Before we entered loop but `rx.wait()` should unblock when sender is dropped and return `Ok(Ready(None))`